src/tracing/esp32_trace/include/matter/tracing/insights_sys_stats.cpp

```diff
@@ -0,0 +1,158 @@
+        if (err != ESP_OK)
+        {
+            ESP_LOGE(kTag, "Failed to add the metric:%s, err:%d", instance->mLabels[i], err);
+        }
+    }
+    DeviceLayer::SystemLayer().StartTimer(instance->GetSamplingInterval(), SamplingHandler, instance);
+}
+
+CHIP_ERROR InsightsSystemMetrics::Unregister()
+{
+    if (!mRegistered)
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    for (int i = 0; i < System::Stats::kNumEntries; i++)
+    {
+        if (mLabels[i] != nullptr)
+        {
+            esp_err_t err = esp_diag_metrics_unregister(mLabels[i]);
+            if (err != ESP_OK)
+            {
+                ESP_LOGE(kTag, "Failed to unregister metric:%s, err:%d", mLabels[i], err);
+            }
+            free(mLabels[i]);
+            mLabels[i] = nullptr;
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+
+void InsightsSystemMetrics::SetSamplingHandler(intptr_t arg)
+{
+    auto instance = static_cast<InsightsSystemMetrics *>(reinterpret_cast<void *>(arg));
+
+    if (instance->mTimeout == System::Clock::kZero)
+    {
+        DeviceLayer::SystemLayer().CancelTimer(SamplingHandler, instance);
+    }
+    else
+    {
+        DeviceLayer::SystemLayer().CancelTimer(SamplingHandler, instance);
+        CHIP_ERROR err = DeviceLayer::SystemLayer().StartTimer(instance->mTimeout, SamplingHandler, instance);
+        if (err != CHIP_NO_ERROR)
+        {
+            ESP_LOGE(kTag, "Failed to start the new timer %" CHIP_ERROR_FORMAT, err.Format());
+        }
+    }
+}
+
+CHIP_ERROR InsightsSystemMetrics::SetSamplingInterval(chip::System::Clock::Timeout aTimeout)
+{
+    if (!mRegistered)
+    {
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    mTimeout = aTimeout;
+    DeviceLayer::PlatformMgr().ScheduleWork(SetSamplingHandler, reinterpret_cast<intptr_t>(this));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR InsightsSystemMetrics::RegisterAndEnable(chip::System::Clock::Timeout aTimeout)
+{
+    if (mRegistered)
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    if (aTimeout == System::Clock::kZero)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    const Label * labels = GetStrings();
+    for (int i = 0; i < System::Stats::kNumEntries; i++)
+    {
+        size_t labelLength = strlen(labels[i]);
+        if (labelLength >= kMaxStringLength)
+        {
+            labelLength = kMaxStringLength;
+            mLabels[i]  = strndup(labels[i], labelLength - 1);
+        }
+        else
+        {
+            mLabels[i] = strndup(labels[i], labelLength);
+        }
+
+        if (mLabels[i] == NULL)
+        {
+            Unregister();
+            return CHIP_ERROR_NO_MEMORY;
+        }
+
+        esp_err_t err = esp_diag_metrics_register(kTag, mLabels[i], labels[i], kPath, ESP_DIAG_DATA_TYPE_UINT);
+        if (err != ESP_OK)
+        {
+            ESP_LOGE(kTag, "Failed to register metric:%s, err:%d", mLabels[i], err);
+            return CHIP_ERROR_INCORRECT_STATE;
```

Comment:
Should we be calling `Unregister()` here since registration has failed for a smoother cleanup?